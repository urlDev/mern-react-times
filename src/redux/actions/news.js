import axios from 'axios';

export const FETCH_POPULAR_SUCCESS = 'FETCH_POPULAR_SUCCESS';
export const FETCH_STORY_SUCCESS = 'FETCH_STORY_SUCCESS';
export const FETCH_NEWS_ERROR = 'FETCH_NEWS_ERROR';
export const CHANGE_HEADER = 'CHANGE_HEADER';

export const fetchPopularSuccess = (news) => ({
    type: FETCH_POPULAR_SUCCESS,
    payload: news,
});

export const fetchStorySuccess = (news) => ({
    type: FETCH_STORY_SUCCESS,
    payload: news,
});

export const changeHeader = (header) => ({
    type: CHANGE_HEADER,
    payload: header,
});

export const fetchNewsError = (error) => ({
    type: FETCH_NEWS_ERROR,
    payload: error,
});

export const fetchTopStories = (topStories) => async(dispatch) => {
    try {
        const response = await axios.get(
            `https://api.nytimes.com/svc/topstories/v2/${topStories}.json?api-key=${process.env.REACT_APP_NEWS_API_KEY}`,
        );
        const data = await response.data;
        return dispatch(fetchStorySuccess(data));
    } catch (error) {
        return dispatch(fetchNewsError(error));
    }
};

export const fetchMostPopular = () => async(dispatch) => {
    try {
        const response = await axios.get(
            `https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=${process.env.REACT_APP_NEWS_API_KEY}`,
        );
        const data = await response.data;
        return dispatch(fetchPopularSuccess(data));
    } catch (error) {
        return dispatch(fetchNewsError(error));
    }
};