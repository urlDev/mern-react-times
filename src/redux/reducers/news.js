import {
  FETCH_NEWS_ERROR,
  FETCH_POPULAR_SUCCESS,
  FETCH_STORY_SUCCESS,
  CHANGE_HEADER,
  CLEAN_STATE,
  CLEAR_ERROR,
  SET_WIDTH,
  OPEN_RESPONSIVE_MENU,
  CLOSE_RESPONSIVE_MENU,
} from "../actions/news";

const initialState = {
  loadingNews: true,
  popular: [],
  story: [],
  errorNews: null,
  header: "home",
  width: window.innerWidth,
  responsiveMenu: false,
};

const newsReducer = (state = initialState, action) => {
  switch (action.type) {
    case CLEAR_ERROR:
      return {
        ...state,
        errorNews: null,
      };
    case FETCH_POPULAR_SUCCESS:
      return {
        ...state,
        loadingNews: false,
        popular: action.payload.results.slice(0, 4),
      };
    case FETCH_STORY_SUCCESS:
      return {
        ...state,
        loadingNews: false,
        story: action.payload.results.slice(0, 3),
      };
    case FETCH_NEWS_ERROR:
      return {
        ...state,
        errorNews: action.payload,
      };
    case CHANGE_HEADER:
      return {
        ...state,
        header: action.payload,
      };
    case SET_WIDTH:
      return {
        ...state,
        width: action.payload,
      };
    case OPEN_RESPONSIVE_MENU:
      return {
        ...state,
        responsiveMenu: true,
      };
    case CLOSE_RESPONSIVE_MENU:
      return {
        ...state,
        responsiveMenu: false,
      };
    default:
      return state;
  }
};

export default newsReducer;
