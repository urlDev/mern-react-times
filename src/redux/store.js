import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));