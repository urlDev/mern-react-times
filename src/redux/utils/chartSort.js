const chartSort = (prevState, newState) => {
    const tempState = [...prevState, [...newState]];

    let sortedState = [...tempState];

    sortedState = sortedState
        .sort((a, b) => a[1] - b[1])
        .map((stock) => stock.slice(0, 1));

    return sortedState;
};

export default chartSort;