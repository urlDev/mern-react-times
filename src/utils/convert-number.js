const convertNumber = (labelValue) => {
  // Nine Zeroes for Billions
  return Math.abs(Number(labelValue)) >= 1.0e9
    ? (Math.abs(Number(labelValue)) / 1.0e9).toFixed(2) + "B"
    : // Six Zeroes for Millions
    Math.abs(Number(labelValue)) >= 1.0e6
    ? (Math.abs(Number(labelValue)) / 1.0e6).toFixed(2) + "M"
    : // Three Zeroes for Thousands
    Math.abs(Number(labelValue)) >= 1.0e3
    ? (Math.abs(Number(labelValue)) / 1.0e3).toFixed(2) + "K"
    : Math.abs(Number(labelValue));
};

export default convertNumber;
