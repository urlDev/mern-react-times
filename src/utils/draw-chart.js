// Reusable function to draw historical price chart in different components
const drawChart = ({ width, reDraw, chartData, svg, d3, index }) => {
  // reDraw => Charts in market details
  // because they are drawn according to screen sizes

  let height;
  let padding;
  let chartWidth;
  let dates;

  // reDraw is for when marketDetails page is shown
  // Chart is constantly redrawn then so thats why I made a separation
  if (reDraw) {
    // changing chart width according to screen width
    chartWidth = width < 1400 ? (width < 700 ? width - 30 : width - 200) : 550;
    height = width < 1400 ? chartWidth * 0.65 : 345;
    padding = 50;
  } else {
    // when its used in market components, width and height
    chartWidth = 100;
    height = 70;
    padding = 10;
  }

  svg.attr("width", chartWidth).attr("height", height);

  if (reDraw) {
    dates = chartData.map((value) => new Date(value.date));
  } else {
    // In market components, there are more than one chart
    // So I get their indexes from the component
    dates = chartData[index].map((value) => new Date(value.date));
  }

  // xScale
  const xAxisScale = d3
    .scaleTime()
    .domain([d3.min(dates), d3.max(dates)])
    .range(
      // Depending if its reDrawn or width changes,
      // In smaller screens, I don't put any padding
      reDraw
        ? width < 500
          ? [0, chartWidth]
          : [padding, chartWidth - padding]
        : [padding, chartWidth - padding]
    );

  // yScale
  const yAxisScale = d3
    .scaleLinear()
    .domain([
      // Chart is drawn from close data on each dataset
      // So in here I am checking where the function is being used
      d3.min(reDraw ? chartData : chartData[index], (data) => data.close),
      d3.max(reDraw ? chartData : chartData[index], (data) => data.close),
    ])
    .range([height - padding, padding]);

  // Charts in market cards don't use axes
  if (reDraw) {
    //   create x Axes
    const xAxes = d3
      .axisBottom(xAxisScale)
      .ticks(width < 700 ? d3.timeDay.every(10) : 5)
      .tickFormat((date) => d3.timeFormat("%b %d, %I:%M")(date));

    // draw x axes
    width > 500 &&
      svg
        .append("g")
        .call(xAxes)
        .attr("transform", `translate(0,${height - padding})`);

    // create y axes
    const yAxes = d3.axisLeft(yAxisScale).tickFormat((d) => `$ ${d}`);

    // draw y axes
    width > 500 &&
      svg.append("g").call(yAxes).attr("transform", `translate(${padding},0)`);
  }

  // create line
  const line = d3
    .line()
    .x((d) => xAxisScale(new Date(d.date)))
    .y((d) => yAxisScale(d.close));

  // draw historical price path
  const path = svg
    .append("path")
    .data(reDraw ? [chartData] : [chartData[index]])
    .style("fill", " none")
    .attr("stroke", reDraw ? "black" : "white")
    .attr("stroke-width", reDraw ? "1.5" : "1.2")
    .attr("d", line);

  // get the length of the path for line transition
  const pathLength = path.node().getTotalLength();

  // transition for the line
  const transitionPath = d3.transition().ease(d3.easeSin).duration(2000);

  // setting pathsLength for strokes attrs and using transition
  // I made it minus pathLength because it was drawing it backwards
  path
    .attr("stroke-dashoffset", -pathLength)
    .attr("stroke-dasharray", pathLength)
    .transition(transitionPath)
    .attr("stroke-dashoffset", 0);

  path.interrupt("stroke-dashoffset");
  path.interrupt("stroke-dasharray");
};

export default drawChart;
