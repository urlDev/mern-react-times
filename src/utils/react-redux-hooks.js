// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
