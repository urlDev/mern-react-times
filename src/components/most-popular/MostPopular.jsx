import React from "react";
import { useSelector } from "../../utils/react-redux-hooks";

import NewsPlaceholder from "../../assets/newsPlaceholder.png";

import {
  MostPopularContainer,
  PopularTitle,
  PopularStoriesContainer,
  StoryContainer,
  StoryTitle,
  ImageContainer,
  SeeAll,
  MostPopularWithHeader,
} from "./MostPopular.styles";

const MostPopular = () => {
  const { popular, width } = useSelector((news) => news.news);

  // Wanted to show random results on each refresh/load
  // but random makes the app loads more than once
  // const random = Math.floor(Math.random() * popular.length);

  return (
    <MostPopularWithHeader>
      <PopularTitle>Most Popular</PopularTitle>
      <MostPopularContainer>
        {popular.length > 0
          ? popular.map((story) => (
              <PopularStoriesContainer
                key={story.id}
                href={story.url}
                target="_blank"
              >
                <StoryContainer>
                  <StoryTitle font="var(--font-header)">
                    {width > 830
                      ? story.title.split(" ").slice(0, 6).join(" ")
                      : story.title}
                    ...
                  </StoryTitle>
                  <StoryTitle color="var(--gray)">
                    {width > 830
                      ? story.abstract.split(" ").slice(0, 10).join(" ")
                      : story.abstract}
                    ...
                  </StoryTitle>
                </StoryContainer>
                <ImageContainer
                  background={
                    story.media[0]
                      ? story.media[0]["media-metadata"][2].url
                      : NewsPlaceholder
                  }
                />
              </PopularStoriesContainer>
            ))
          : null}
        <SeeAll
          popular="43px"
          href="https://www.nytimes.com/trending/?action=click&contentCollection=Africa&module=MostEmailed&pgtype=article&region=Marginalia&src=me&version=Full"
          target="_blank"
        >
          See All
        </SeeAll>
      </MostPopularContainer>
    </MostPopularWithHeader>
  );
};

export default MostPopular;
