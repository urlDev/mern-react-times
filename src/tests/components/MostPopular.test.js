import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import "jest-styled-components";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import MostPopular from "components/most-popular/MostPopular";
import { popularWithoutMedia, width } from "tests/fixtures/news";
import {
  ImageContainer,
  StoryTitle,
} from "components/most-popular/MostPopular.styles";

describe("Testing MostPopular component", () => {
  let store;
  let wrapper;

  beforeEach(() => {
    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  describe("Testing the component with popular news fetched", () => {
    beforeEach(() => {
      store = mockStore({
        popular: popularWithoutMedia,
        width,
      });

      wrapper = mount(<MostPopular store={store} />);
    });

    test("Should render with popular news", () => {
      expect(wrapper).toHaveLength(1);
    });
    test("Should match the snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
    test("Should get the state", () => {
      const state = store.getState();

      expect(state).toEqual({ popular: popularWithoutMedia, width });
    });
  });
  describe("Testing the component without any news state", () => {
    beforeEach(() => {
      store = mockStore({
        popular: [],
      });

      wrapper = shallow(<MostPopular store={store} />);
    });

    test("Should render component without news", () => {
      expect(wrapper).toHaveLength(1);
    });
    test("Should match the snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  test("Should apply styles to ImageContainer according to url prop", () => {
    wrapper = renderer
      .create(<ImageContainer background="/imageUrl" />)
      .toJSON();
    // Passing background url prop
    expect(wrapper).toHaveStyleRule("background", "url(/imageUrl) no-repeat");
  });
  test("Should apply styles to StoryTitle according to default", () => {
    wrapper = renderer.create(<StoryTitle />).toJSON();
    // Default values in styles
    expect(wrapper).toHaveStyleRule("font-family", "var(--font-text)");
    expect(wrapper).toHaveStyleRule("color", "black");
  });
  test("Should apply styles to StoryTitle according to passed props", () => {
    wrapper = renderer
      .create(<StoryTitle font="var(--font-header)" color="white" />)
      .toJSON();
    // Passed props to component
    expect(wrapper).toHaveStyleRule("font-family", "var(--font-header)");
    expect(wrapper).toHaveStyleRule("color", "white");
  });
});
