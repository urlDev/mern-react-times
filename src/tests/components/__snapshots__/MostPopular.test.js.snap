// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MostPopular component Testing the component with popular news fetched Should match the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 55px;
  padding: 0 0 0 40px;
}

.c1 {
  font-family: var(--font-header);
  font-size: var(--size-category-selected);
  font-weight: 900;
  margin-top: 0;
  margin-bottom: 20px;
}

.c3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #707070;
  font-size: var(--size-sub-menu);
  font-family: var(--font-links);
  font-weight: bold;
  margin-bottom: 20px;
  cursor: pointer;
  display: inline-block;
  position: relative;
  -webkit-transition: color 0.3s ease-in;
  transition: color 0.3s ease-in;
}

.c3:hover {
  color: black;
}

.c3:after {
  content: "";
  position: absolute;
  bottom: -7px;
  left: 0;
  width: 43px;
  background: var(--gray);
  height: 2px;
}

@media (max-width:1400px) {
  .c2 {
    width: 100%;
    min-width: 300px;
  }
}

@media (max-width:830px) {
  .c2 {
    display: grid;
    grid-template-columns: repeat(2,1fr);
    grid-column-gap: 30px;
    grid-row-gap: 50px;
  }
}

@media (max-width:600px) {
  .c2 {
    grid-template-columns: 1fr;
    min-width: 100%;
  }
}

@media (max-width:830px) {
  .c0 {
    margin: 0;
    padding: 0;
  }
}

<MostPopular
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.h1>
        <h1
          className="c1"
        >
          Most Popular
        </h1>
      </styled.h1>
      <styled.div>
        <div
          className="c2"
        >
          <styled.a
            href="https://www.nytimes.com/trending/?action=click&contentCollection=Africa&module=MostEmailed&pgtype=article&region=Marginalia&src=me&version=Full"
            popular="43px"
            target="_blank"
          >
            <a
              className="c3"
              href="https://www.nytimes.com/trending/?action=click&contentCollection=Africa&module=MostEmailed&pgtype=article&region=Marginalia&src=me&version=Full"
              target="_blank"
            >
              See All
            </a>
          </styled.a>
        </div>
      </styled.div>
    </div>
  </styled.div>
</MostPopular>
`;

exports[`Testing MostPopular component Testing the component without any news state Should match the snapshot 1`] = `
<styled.div>
  <styled.h1>
    Most Popular
  </styled.h1>
  <styled.div>
    <styled.a
      href="https://www.nytimes.com/trending/?action=click&contentCollection=Africa&module=MostEmailed&pgtype=article&region=Marginalia&src=me&version=Full"
      popular="43px"
      target="_blank"
    >
      See All
    </styled.a>
  </styled.div>
</styled.div>
`;
